storage/src/file_storage.rs

```diff
@@ -0,0 +1,200 @@
+        let mut file = fs::File::open(full_path).await.map_err(|_| {
+            StorageErrorKind::DoesNotExist.with_error(anyhow::anyhow!(
+                "Failed to find dest_path {}",
+                self.uri(path)
+            ))
+        })?;
+        file.seek(SeekFrom::Start(range.start as u64)).await?;
+        let mut content_bytes = vec![0u8; range.len()];
+        file.read_exact(&mut content_bytes).await?;
+        Ok(content_bytes)
+    }
+
+    async fn delete(&self, path: &Path) -> StorageResult<()> {
+        let full_path = self.root.join(path);
+        fs::remove_file(full_path).await?;
+        Ok(())
+    }
+
+    async fn get_all(&self, path: &Path) -> StorageResult<Vec<u8>> {
+        let full_path = self.root.join(path);
+        let content_bytes = fs::read(full_path).await?;
+        Ok(content_bytes)
+    }
+
+    fn uri(&self) -> String {
+        format!("file://{}", self.root.to_string_lossy())
+    }
+}
+
+/// A File storage resolver
+#[derive(Debug, Clone)]
+pub struct FileStorageFactory {
+    protocol: String,
+}
+
+impl Default for FileStorageFactory {
+    fn default() -> Self {
+        FileStorageFactory {
+            protocol: "file".to_string(),
+        }
+    }
+}
+
+impl StorageFactory for FileStorageFactory {
+    fn protocol(&self) -> String {
+        self.protocol.clone()
+    }
+
+    fn resolve(&self, uri: &str) -> StorageResult<Arc<dyn Storage>> {
+        let pattern = format!("{}://", self.protocol);
+        if !uri.starts_with(pattern.as_str()) {
+            let err_msg = anyhow::anyhow!(
+                "{:?} is an invalid file storage uri. Only file:// is accepted.",
+                uri
+            );
+            return Err(StorageErrorKind::DoesNotExist.with_error(err_msg));
+        }
+
+        let storage = FileStorage::from_uri(uri)?;
+        Ok(Arc::new(storage))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tempfile::tempdir;
+
+    use super::*;
+    use crate::tests::storage_test_suite;
+
+    #[tokio::test]
+    async fn test_storage() -> anyhow::Result<()> {
+        let path_root = tempdir()?;
+        let mut file_storage = FileStorage::new(path_root.into_path());
+        storage_test_suite(&mut file_storage).await?;
+        Ok(())
+    }
+
+    #[test]
+    fn test_file_storage_factory() -> anyhow::Result<()> {
+        let file_storage_factory = FileStorageFactory::default();
+        let storage = file_storage_factory.resolve("file://foo/bar")?;
+        assert_eq!(storage.uri().as_str(), "file://foo/bar");
+
+        let err = file_storage_factory
+            .resolve("test://foo/bar")
+            .err()
+            .unwrap();
+        assert_eq!(err.kind(), StorageErrorKind::DoesNotExist);
+
+        let err = file_storage_factory.resolve("test://").err().unwrap();
+        assert_eq!(err.kind(), StorageErrorKind::DoesNotExist);
+        Ok(())
+    }
+
+    // #[tokio::test]
+    // async fn test_file_storage_builder() -> anyhow::Result<()> {
```

Comment:
no need for a builder.

Reply:
I was still cleaning it. so I just pushed my latest proofread